### ISSUE TITLE
way to force logstash to process tmp files?

### ISSUE THREAD

**Issue by reschex** [Azure/logstash-output-kusto#11]
Hi

What determines when logstash sends the temp files to kusto? 
Is there a way to force the processing of all existing temp files?

I've got some integration tests that verify that some records show up in kusto as end of the pipeline, and right now I'm still playing with increasing the the wait time to give logstash a chance to close the files and upload them, but that feels very clunky.. I'd rather force the plugin to process the files or reduce the time it takes for the plugin to close existing files and process them.

Thanks!

Roland

**Comment by ncbrown1** [comment]
Current Case:
I've found that the temp file gets processed/ingested once the file rotates (e.g. to the next minute). 

If you can find a way to force-rotate files at a pre-defined interval, that would take care of your issue without needing updates to logstash-output-kusto. That's my take, at least.

**Comment by reschex** [comment]
Thanks both!
That makes sense. What confused me was that nothing happened after a minute. 
The scenario is: a jenkins pipeline will deploy a logstash container with the kusto plugin into k8s - an integration test container is then started to run through some behave features/scenarios (when data with value X is send, it shows up in kusto db X etc). The problem was/is that the test sends 3 events, which opens 3 tmp files (as expected) but then nothing happens. I've had a sleep timer for 5mins sat as a step after sending data and watched the files staying open.

So I've added a step after sending the events to wait 60sec, then send another 3 events and then wait another 60sec to give kusto a chance to ingest. This does have the desired effect of the original 3 files being closed and deleted (presumably send for ingest) and 3 new files being opened with a new timestamp - I'm happy to use that method (I need to figure out why the events don't show up in kusto,but thats a seperate issue for slack my thinks :) )

That does leave the question though.. when will the kusto plugin process files that are open when it doesn't receive new events? i.e. the 3 files my test opens sat untouched for at least 5mins without being processed - is there a timeout for that where the plugin will eventually process everything it has open?

Thanks for your help!

Roland

**Comment by tamirkamara** [comment]
Currently the plugin processing triggers are depended on events coming is This means that if no event comes in, the plugin will not close/send the temp files it has.

**Comment by reschex** [comment]
Thanks for the confirmation - that means in a worst case scenario, I might lose 1 minute worth of logs - i.e. if an event source stops producing 1sec before the log rotate, the logs either not get ingested at all in case of logstash being forcefully terminated some time down the line or with a long delay either only when the event source resumes sending or logstash is properly restarted.

Feels like it needs a parameter to control how long the temp files are being kept open with no activity before they are being closed and uploaded - ideally something that can be set in the output config ;)

**Comment by binoysankar** [comment]
Hi Team,
I have similar issue. I am using Kusto plugins to push from Azure AKS to ADX and yesterday my app_id and app_key got expired and the Logstash pod kept all the tmp files from uploading to ADX and the logs had error. I renewed the service principal and updated in ConfigMap and I want these tmp files from yesterday to push it to ADX.
I backed up the files to my local from the pod and tried to kill the Logstash container so that it pick the old files. but it only picks the new files from today.

how can I get my old files in /tmp/kusto/ to be pushed to ADX?

**Comment by ag-ramachandran** [comment]
Hello @binoysankar 

What is ther version of the plugin you are using ? Depending on the version of the flag, these files should get picked up

```recovery	If set to true (default), plugin will attempt to resend pre-existing temp files found in the path upon startup	```

You can look at "touch"ing these files and they should get picked up for processing from the same path which you have configured.

**Comment by ag-ramachandran** [comment]
Thanks for the reply @binoysankar , usually the logstash plugin would have a version. There is also  list on the logstash plugins that should give you the version

Now that the files are getting pushed, the issue is no longer a blocker.

**Comment by binoysankar** [comment]
> Thanks for the reply @binoysankar , usually the logstash plugin would have a version. There is also list on the logstash plugins that should give you the version
> 
> Now that the files are getting pushed, the issue is no longer a blocker.

I have pulled in the latest plugin and did a redeployment of Logstash. My doubt is if for some reason which I mentioned above Logstash pod is not able to push the logs to ADX the logs will accumulate in ```/tmp/kusto/```. Do I need to kill the container for this to work. In doc it says during start up the recovery process will fetch the old tmp files and push it. Its not very clear to me what should be done when something happens. 

What I obervered is when the pod restarts or recreate the container the /tmp/kusto directory is recreate and old files are lost.
